game-core/src/main/java/games/strategy/engine/data/UnitHitsChange.java

```diff
@@ -10,35 +9,42 @@
 
   private final IntegerMap<Unit> hits;
   private final IntegerMap<Unit> undoHits;
+  private final Collection<Territory> territoriesToNotify;
 
-  private UnitHitsChange(final IntegerMap<Unit> hits, final IntegerMap<Unit> undoHits) {
+  private UnitHitsChange(
+      final IntegerMap<Unit> hits,
+      final IntegerMap<Unit> undoHits,
+      final Collection<Territory> territoriesToNotify) {
     this.hits = hits;
     this.undoHits = undoHits;
+    this.territoriesToNotify = territoriesToNotify;
   }
 
-  public UnitHitsChange(final IntegerMap<Unit> hits) {
-    this.hits = new IntegerMap<>(hits);
-    undoHits = new IntegerMap<>();
-    for (final Unit item : this.hits.keySet()) {
+  public UnitHitsChange(
+      final IntegerMap<Unit> hits, final Collection<Territory> territoriesToNotify) {
+    this(new IntegerMap<>(hits), undoHits(hits), territoriesToNotify);
+  }
+
+  private static IntegerMap<Unit> undoHits(final IntegerMap<Unit> hits) {
+    final var undoHits = new IntegerMap<Unit>();
+    for (final Unit item : hits.keySet()) {
       undoHits.put(item, item.getHits());
     }
+    return undoHits;
   }
 
   @Override
   protected void perform(final GameData data) {
     for (final Unit item : hits.keySet()) {
       item.setHits(hits.getInt(item));
     }
-    final Set<Unit> units = hits.keySet();
-    for (final Territory element : data.getMap().getTerritories()) {
-      if (!Collections.disjoint(element.getUnits(), units)) {
```

Comment:
Is this disjoint check just not needed in the end?

Reply:
That's right. The old code was using it to determine if the affected units are in the territory in question. Now, we track that directly, so this is removed. (This call was what was so slow in the first place.)